wire/messages.go

```diff
@@ -0,0 +1,175 @@
+package wire
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/routing"
+
+	logging "github.com/ipfs/go-log"
+	record "github.com/libp2p/go-libp2p-record"
+	recpb "github.com/libp2p/go-libp2p-record/pb"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/libp2p/go-libp2p-kad-dht/internal"
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+)
+
+var logger = logging.Logger("dht")
+
+// ProtocolMessenger can be used for sending DHT messages to peers and processing their responses.
+// This decouples the wire protocol format from both the DHT protocol implementation and from the implementation of the
+// routing.Routing interface.
+//
+// TODO: This is still strongly coupled with the existing implementation of what happens when a peer actually sends a
+// message on the wire (e.g. reusing streams, reusing connections, metrics tracking, etc.).
+type ProtocolMessenger struct {
+	m         MessageSender
+	validator record.Validator
+}
+
+type ProtocolMessengerOption func(*ProtocolMessenger) error
+
+func WithValidator(validator record.Validator) ProtocolMessengerOption {
```

Comment:
The only option right now is the validator. I'm not currently convinced that the validator belongs here instead of in the user of the ProtocolMessenger.

Having an option parameter seems like it could be useful anyway though so this issue seems pretty low stakes (i.e. returning err from NewProtocolMessenger is annoying, but might be worth doing to save ourselves from breaking changes later).

Thoughts?

Reply:
Leaving this for now, can always change later